### PR TITLE
Convert diff panel component to React and various fixes

### LINKED ISSUE #174: Simultaneously selecting changes to publish and undo behaves unexpectedly

### PR DESCRIPTION
**Pull request by EvanOxfeld**
- [x] @sfeng88 @rygim @jharwig 
- [x] @srfarley @joeferner
- [x] @EvanOxfeld @joeybrk372
- [ ] @mwizeman @dsingley

Converts diff panel from a flight component to a single React component as the first step in improving the diff panel's state management, which had become challenging to get correct, and performance. For this initial pull request 1) the diff panel's EJS templates were adapted into JSX with as little changed as possible with the goal of rendering the same DOM elements down to the (potentially unused) data attributes and 2) moved state management outside of the DOM. Future pull requests will break down the diff component further for increased performance and move state management into React to use its more fleshed out component lifecycle.

Additionally adds a living React (and eventually ES6) style guide put together with the help of @jharwig.

Defects fixed:
* Fixes #174
* Fixes header buttons flashing for a few seconds on receiving a `diffsChanged` event
* Fixes action buttons reappearing while publishing or undoing
* Fixes vertices and edges selected on the graph not appearing as selected in the diff panel when first opened
* Fixes an undo button click on a property marked as undo causing the vertex or edge to no longer be selected for publish.
* Fixes strikethough of a modified visibility in the diff panel

### CODE REVIEW

File: web/war/src/main/webapp/README.md [hunk 1]
```diff
@@ -1,3 +1,161 @@
...
+
+## React
+
+### Basic Rules
+
+  - Prefer including only one React component per file.
+
+### `React.createClass` vs stateless function
+
+  - If the component maintains internal state, create the React class using `React.createClass`. Avoid `class extends React.Component`. eslint: [`react/prefer-es6-class#never-mode`](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/prefer-es6-class.md#never-mode)
```
Comment (joeferner) [inline web/war/src/main/webapp/README.md]: Any reasoning behind this?

File: web/war/src/main/webapp/js/require.config.js [hunk 2]
```diff
@@ -98,10 +103,17 @@
...
             'rangy-cssclassapplier': { deps: ['rangy-core'] },
             'rangy-serializer': { deps: ['rangy-core'] },
+            'react': { exports: 'React' },
             'jquery-scrollstop': { exports: 'jQuery', deps: ['jquery'] },
             'underscore': { exports: '_' },
             'videojs': { exports: 'videojs' }
         },
+        config: {
+            babel: {
+                sourceMaps: 'inline',
```
Comment (joeferner) [inline web/war/src/main/webapp/js/require.config.js]: In a prod build is this turned off somehow?

File: web/war/src/main/webapp/js/workspaces/diff/DiffPanel.jsx [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+
+                    {publishCount === 0 && undoCount === 0 ? (
+                        <span>{ i18n('workspaces.diff.header.unpublished_changes') }</span>
+                    ) : null}
+                  </h1>
+              </div>
+            );
+        },
+
+        renderHeaderActions: function(publishingAll, undoingAll) {
```
Comment (joeferner) [inline web/war/src/main/webapp/js/workspaces/diff/DiffPanel.jsx]: Should this be a separate component?

File: web/war/src/main/webapp/js/workspaces/diff/DiffPanel.jsx [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+            const publishingAll = publishCount === flatDiffs.length;
+            const undoingAll = undoCount === flatDiffs.length;
+
+            return (
+              <div className="diff-header">
+                  {this.renderHeaderActions(publishingAll, undoingAll)}
+
+                  <h1 className="header">
+                    {publishing || publishCount > 0 ? (
+                        <button className={
```
Comment (joeferner) [inline web/war/src/main/webapp/js/workspaces/diff/DiffPanel.jsx]: How are these buttons different from the ones in `renderHeaderActions`?

File: web/war/src/main/webapp/js/workspaces/diff/DiffPanel.jsx [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+                                {diffs.map(this.renderDiff)}
+                            </tbody>
+                        </table>
+                    </div>
+                </div>
+            );
+        },
+
+        //TODO add
+        propTypes: {
```
Comment (joeferner) [inline web/war/src/main/webapp/js/workspaces/diff/DiffPanel.jsx]: lame

File: web/war/src/main/webapp/README.md (hunk 1 above)
Comment (EvanOxfeld) [inline web/war/src/main/webapp/README.md]: A couple main reasons:

* The `class` keyword is just syntactic sugar, potentially distracting from prototypal inheritance and its differences with classical inheritance.
* Defining React components as ES6 classes only binds `this` to the instance in React API methods. See https://facebook.github.io/react/docs/reusable-components.html#no-autobinding for more details.

File: web/war/src/main/webapp/js/require.config.js (hunk 2 above)
Comment (EvanOxfeld) [inline web/war/src/main/webapp/js/require.config.js]: I'm ok with leaving this option on in production until something like webpack is in place.

File: web/war/src/main/webapp/js/workspaces/diff/DiffPanel.jsx (hunk 3 above)
Comment (EvanOxfeld) [inline web/war/src/main/webapp/js/workspaces/diff/DiffPanel.jsx]: Yes it should. `DiffPanel` is rather monolithic as React components go. That said, I see breaking up the component as outside of the scope of this pull request, although it's a task I'd like to tackle in the near future to improve performance when the list of diffs is large.

File: web/war/src/main/webapp/js/workspaces/diff/DiffPanel.jsx (hunk 4 above)
Comment (EvanOxfeld) [inline web/war/src/main/webapp/js/workspaces/diff/DiffPanel.jsx]: Yeah this is a little confusing and the header could benefit from better function/component names. The buttons starting on this line are the larger buttons that apply (publish or undo) the respective marked diffs: screen shot 2016-05-11 at 3 38 16 pm

`renderHeaderActions` is a direct conversion of `actions.ejs` to React and is responsible for the select all publish and undo buttons: screen shot 2016-05-11 at 3 39 13 pm

File: web/war/src/main/webapp/js/workspaces/diff/DiffPanel.jsx (hunk 5 above)
Comment (EvanOxfeld) [inline web/war/src/main/webapp/js/workspaces/diff/DiffPanel.jsx]: Good catch -- that was supposed to be in this pull request. Fixed.

**Pull request merged**